disktable.go

```diff
@@ -0,0 +1,650 @@
+		if err != nil {
+			return err
+		}
+		if !bytes.Equal(item.Key(), key) {
+			panic("known key search came up with different key: HUGE PROBLEM, HUGE!!!!!!")
+		}
+		value, err := item.ValueCopy(nil)
+		if err != nil {
+			return fmt.Errorf("problem getting value of key(%s): %w", string(key), err)
+		}
+		result = lookupResult{i: endian.Uint64(value)}
+		return nil
+	}
+
+	if err := index.db.View(f); err != nil {
+		result.err = err
+	}
+	return result
+
+}
+
+// multiMatch matches all keys in an index that allows duplicates. key must have been finalKey()'d.
+func (t *Table) multiMatch(key []byte, index *Index, ch chan lookupResult) {
+	defer close(ch)
+
+	keyData, err := getPrefixSuffix(key)
+	if err != nil {
+		ch <- lookupResult{err: err}
+		return
+	}
+
+	tempBuff := make([]byte, 8)
+	err = index.db.View(
+		func(txn *badger.Txn) error {
+			it := txn.NewIterator(badger.DefaultIteratorOptions)
+			defer it.Close()
+
+			for it.Seek(keyData.prefix); it.ValidForPrefix(keyData.prefix); it.Next() {
+				item := it.Item()
+
+				foundKeyData, err := getPrefixSuffix(item.Key())
+				if err != nil {
+					return fmt.Errorf("multiMatch of index(%s) had bad key(%v): %w", index.Name, item.Key(), err)
+				}
+				// It is possible to have a prefix match for something not exactly the same.
+				// But we are looking for exact prefix matches followed by a number followed by
+				// our key size. By matching prefix and key size, we can be sure of an exact match.
+				// If not, throw out the data.
+				if len(foundKeyData.prefix) != len(keyData.prefix) {
+					continue
+				}
+				b, err := item.ValueCopy(tempBuff)
+				if err != nil {
+					return err
+				}
+				ch <- lookupResult{i: endian.Uint64(b)}
+			}
+			return nil
+		},
+	)
+	if err != nil {
+		ch <- lookupResult{err: err}
+	}
+}
+
+func (t *Table) FetchAll(ctx context.Context) (chan Result, error) {
+	stream := t.primary.db.NewStream()
+	stream.NumGo = 16
+	stream.LogPrefix = "FetchAll.Streaming"
+
+	ch := make(chan Result, 1)
+	stream.Send = func(buf *z.Buffer) error {
+		list, err := badger.BufferToKVList(buf)
+		if err != nil {
+			return err
+		}
+
+		for _, kv := range list.Kv {
+			// TODO(jdoak): This doesn't seem to matter.
+			v := make([]byte, len(kv.Value))
+			copy(v, kv.Value)
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case ch <- Result{Value: v}:
+			}
+		}
+		return nil
+	}
+
+	go func() {
+		defer close(ch)
+		if err := stream.Orchestrate(context.Background()); err != nil {
+			ch <- Result{Err: err}
+			return
+		}
+	}()
+
+	return ch, nil
+}
+
+// ByteSlice2String coverts bs to a string. It is no longer safe to use bs after this.
+// This prevents having to make a copy of bs.
+func ByteSlice2String(bs []byte) string {
+	if len(bs) == 0 {
+		return ""
+	}
+	return *(*string)(unsafe.Pointer(&bs))
+}
+
+// UnsafeGetBytes retrieves the underlying []byte held in string "s" without doing
+// a copy. Do not modify the []byte or suffer the consequences.
```

Comment:
just for my own curiosity, is the behavior undefined if you do modify it?

Reply:
It would depend on how it is being used.  In a single thread, I think you could make the argument that you can look at the code and see what will happen.  If you are changing the []byte in a separate goroutine, then it would become undefined because race conditions in go are undefined, where a normal string cannot have a race condition.

Reply:
oh, I assumed you meant "don't change the underlying `[]byte` because you'll mess up the given `string`". the race condition part makes sense

Reply:
It will mess up the given string, which certainly violates rules.  But I think the cause and effect for that can be traced, so I'm not sure I'd say it was undefined until you do it with a race condition.